ReactiveCocoa/NSObject+Intercepting.swift

```diff
@@ -6,30 +6,419 @@ extension Reactive where Base: NSObject {
+		let forwardInvocation = unsafeBitCast(impl, to: SuperForwardInvocation.self)
+		forwardInvocation(objectRef, ObjCSelector.forwardInvocation, invocation)
+	}
+
+	_ = class_replaceMethod(realClass,
+	                        ObjCSelector.forwardInvocation,
+	                        imp_implementationWithBlock(newForwardInvocation as Any),
+	                        ObjCMethodEncoding.forwardInvocation)
+}
+
+/// Swizzle `realClass` to accelerate the method signature retrieval, using a
+/// signature cache that covers all known intercepted selectors of `realClass`.
+///
+/// - parameters:
+///   - realClass: The runtime subclass to be swizzled.
+///   - signatureCache: The method signature cache.
+private func setupMethodSignatureCaching(_ realClass: AnyClass, _ signatureCache: SignatureCache) {
+	let perceivedClass: AnyClass = class_getSuperclass(realClass)
+
+	let newMethodSignatureForSelector: @convention(block) (Unmanaged<NSObject>, Selector) -> AnyObject? = { objectRef, selector in
+		if let signature = signatureCache[selector] {
+			return signature
+		}
+
+		typealias SuperMethodSignatureForSelector = @convention(c) (Unmanaged<NSObject>, Selector, Selector) -> AnyObject?
+		let impl = class_getMethodImplementation(perceivedClass, ObjCSelector.methodSignatureForSelector)
+		let methodSignatureForSelector = unsafeBitCast(impl, to: SuperMethodSignatureForSelector.self)
+		return methodSignatureForSelector(objectRef, ObjCSelector.methodSignatureForSelector, selector)
 	}
+
+	_ = class_replaceMethod(realClass,
+	                        ObjCSelector.methodSignatureForSelector,
+	                        imp_implementationWithBlock(newMethodSignatureForSelector as Any),
+	                        ObjCMethodEncoding.methodSignatureForSelector)
+}
+
+/// The state of an intercepted method specific to an instance.
+private final class InterceptingState {
+	let (signal, observer) = Signal<AnyObject, NoError>.pipe()
+
+	/// Initialize a state specific to an instance.
+	///
+	/// - parameters:
+	///   - lifetime: The lifetime of the instance.
+	init(lifetime: Lifetime) {
+		lifetime.ended.observeCompleted(observer.sendCompleted)
+	}
+}
+
+private final class SelectorCache {
+	private var map: [Selector: (main: Selector, interop: Selector)] = [:]
+
+	init() {}
+
+	/// Cache the aliases of the specified selector in the cache.
+	///
+	/// - warning: Any invocation of this method must be synchronized against the
+	///            runtime subclass.
+	@discardableResult
+	func allocate(_ selector: Selector) -> (main: Selector, interop: Selector) {
+		if let pair = map[selector] {
+			return pair
+		}
+
+		var copy = map
+		let aliases = (selector.alias, selector.interopAlias)
+		copy[selector] = aliases
+		map = copy
+
+		return aliases
+	}
+
+	/// Get the alias of the specified selector.
+	///
+	/// - parameters:
+	///   - selector: The selector alias.
+	subscript(main selector: Selector) -> Selector {
```

Comment:
I don't love the use of labeled subscripts for these. I'd personally rather use a method or have a single subscript that returns a `(main: Selector, alias: Selector)`.